document/src/vespa/document/fieldvalue/structuredfieldvalue.cpp

```diff
@@ -87,6 +90,78 @@ StructuredFieldValue::onGetNestedFieldValue(PathRange nested) const
     return fv;
 }
 
+namespace {
+    using ValuePair = std::pair<fieldvalue::ModificationStatus, FieldValue::UP>;
+    using Cache = vespalib::hash_map<Field, ValuePair>;
+}
+
+class StructuredCache {
+public:
+    void remove(const Field & field) {
+        ValuePair & entry = _cache[field];
+        entry.first = ModificationStatus::REMOVED;
+        entry.second.reset();
+    }
+    Cache::iterator find(const Field & field) {
+        return _cache.find(field);
+    }
+    void set(const Field & field, FieldValue::UP value, ModificationStatus status) {
+        ValuePair & entry = _cache[field];
+        entry.first = status;
+        entry.second = std::move(value);
+    }
+    Cache::iterator begin() { return _cache.begin(); }
+    Cache::iterator end() { return _cache.end(); }
+private:
+    Cache _cache;
+};
+
+
+void
+StructuredFieldValue::remove(const Field& field) {
+    if (_cache) {
+        _cache->remove(field);
+    } else {
+        removeFieldValue(field);
+    }
+}
+
+void
+StructuredFieldValue::updateValue(const Field & field, FieldValue::UP value) const {
+    if (_cache) {
+        _cache->set(field, std::move(value), ModificationStatus::MODIFIED);
+    } else {
+        const_cast<StructuredFieldValue&>(*this).setFieldValue(field, std::move(value));
+    }
+}
+
+void
+StructuredFieldValue::returnValue(const Field & field, FieldValue::UP value) const {
+    if (_cache) {
+        _cache->set(field, std::move(value), ModificationStatus::NOT_MODIFIED);
+    }
+}
```

Comment:
Just to make sure my understanding is correct: `returnValue` is only used in a context where the value has been "borrowed" from the underlying cache, and it only makes sense to reinsert it if the cache is active. Otherwise, the value returned from the struct will be a clone of the field value and has nowhere to be returned to.